yarpl/include/yarpl/flowable/Subscriber.h

```diff
@@ -15,40 +15,119 @@ namespace flowable {
 template <typename T>
 class Subscriber : public virtual Refcounted, public yarpl::enable_get_ref {
  public:
+  virtual void onSubscribe(Reference<Subscription>) = 0;
+  virtual void onComplete() = 0;
+  virtual void onError(folly::exception_wrapper) = 0;
+  virtual void onNext(T) = 0;
+};
+
+// codemod all things that inherit from Subscriber to inherit from
+// InternalSubscriber
+template <typename T>
+class InternalSubscriber : public Subscriber<T> {
+ public:
+  // Note: If any of the following methods is overridden in a subclass, the new
+  // methods SHOULD ensure that these are invoked as well.
+  void onSubscribe(Reference<Subscription> subscription) override {
+    DCHECK(subscription);
+    CHECK(!subscription_);
+    subscription_ = std::move(subscription);
+  }
+
+  // No further calls to the subscription after this method is invoked.
+  void onComplete() override {
+    DCHECK(subscription_) << "Calling onComplete() without a subscription";
+    subscription_.reset();
+  }
+
+  // No further calls to the subscription after this method is invoked.
+  void onError(folly::exception_wrapper) override {
+    DCHECK(subscription_) << "Calling onError() without a subscription";
+    subscription_.reset();
+  }
+
+ protected:
+  Reference<Subscription> subscription() {
```

Comment:
What if we make this function `pure virtual`? Can it help getting rid of the `subscription_` data member in the parent class?